File: MAP/Lab6/src/Model/File/OpenFile.java

package Model.File;
import Model.Statement.*;
import Model.*;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class OpenFile implements Statement
{
    private String fileName;
    private String varName;
    public OpenFile(String fl, String vn)
    {
        fileName = fl;
        varName = vn;
    }

    public PrgState execute(PrgState state)
    {
        if(!isOpen(state))
            try
            {
                BufferedReader br = new BufferedReader(new FileReader(fileName));
                FileData fd = new FileData(fileName, br);
                int id = IDGenerator.generate_ID();
                state.getFileTable().add(id, fd);
                IDictionary<String, Integer> dict = state.getSymbolT();
                if(dict.contains(varName))
                {
                    dict.update(varName, id);
                }
                else dict.add(varName, id);
            }
            catch(IOException | NullPointerException ex)
            {
                System.out.println(ex.toString());
            }
        return state;
    }

    private boolean isOpen(PrgState prg)
    {
        for(FileData crt : prg.getFileTable().getValues())
            if(crt.getFileName().equals(fileName))
                return true;
        return false;
    }

    @Override
    public String toString() {
        return "OpenFile{" + "fileName='" + fileName + '\'' + ", varName='" + varName + '\'' + '}';
    }
}

File: MAP/LAB7-MAP/LAB7-MAP/Files/OpenFile.cs
using MAP_Lab7.Statements;
using System;
using System.IO;
using MAP_Lab7.Exceptions;

namespace MAP_Lab7.Files
{
    public class OpenFile : Statement
    {
        private string varName;
        public string fileName;
        private static int ID = 0;

        public OpenFile(string vn, string fn)
        {
            varName = vn;
            fileName = fn;
        }

        public PrgState Execute(PrgState state)
        {
            foreach (FileData i in state.FileTable.AllValues())
                if (i.fileName == fileName)
                    throw new FileException("File is already opened !");

            if (!File.Exists(fileName))
                throw new FileException("File does not exist !");

            StreamReader sw = new StreamReader(fileName, true);
            state.FileTable.Add(ID, new FileData(fileName, sw));
            state.Dict.Update(varName, ID++);

            return state;

        }

        public override string ToString()
        {
            return "Open(" + fileName + ")";

        }
    }
}
